src/telegrambot_lib/http.clj

```diff
@@ -80,6 +80,13 @@
   ([this path content]
    (let [url (gen-url this path)
          req {:body (json/generate-string content)
-              :content-type :json}
-         resp (client :post url req)]
-     (parse-resp resp))))
+              :content-type :json}]
+     (try
+       (let [resp (client :post url req)]
+         (parse-resp resp))
+       (catch Throwable ex
+         (if-let [body (parse-resp (ex-data ex))]
+           ;; matches the async version behavior
+           (assoc body :error ex)
+           ;; rethrows the client code caused ex
+           (throw ex)))))))
```

Comment:
I think that we should avoid re-throwing exceptions here in order to consistently return a map.

If it is unable to create ex-data, then maybe create it with ex-info.

Let me know what you think.

Reply:
Well, I've designed the code this way to separate to possible exceptional situations:
1. An exception coming from the `clj-http` as a result of an unsuccessful Bot API request, e.g. when one calls a method w/o a required parameter, — which should be totally fine, since this is _an expected behavior_. Usually, we do not want an exception here. What we really need is to keep an `ex-data` on a further path to the client code, which itself will decide what to do next (I've mentioned 3 possible reactions in the PR description).
2. An exception coming from elsewhere — an _unexpected_ one, e.g. `SocketException` signaling network issues.

We could distinguish these two situations by checking the `ex-data` of the original exception — if one is here and we are able to retrieve a response body from it, then we are dealing with a legit case (1), not smth. we hope not to happen (2).

Still, there are some issues with this code I was able to spot thanks to you:
1. The case (2) is incorrectly described as the "the client code caused ex" — this is not the root cause and this misleading comment needs to be re-written or thrown away completely (I'd prefer to move the description of these behaviors to the `docstring` of the multi-method + to the library README file);
2. The case (2) is abcent in the async version of the `request` method — and this is a design flaw again which misaligns the two implementations, uh oh... In asynchronous code, where callbacks can/are be processed on a different thread, we, as a user of the library, really have little use in rethrowing exceptions.

Ok, let's stick to passing all exceptions the same way — within an `:error ex` entry of the map. And the convention will be:
- to pass it along with the response map in case the request was unsuccessful in terms of the Bot API (1),
- to return just `{:error ex}` in all other situations when we're unable to retrieve a response body (2).

The client code will have to distinguish and handle all these situations on its own then. Because this is not the library concern — to deal with such problems. The lib should just behave in a fail-fast manner and indicate a failure somehow, without silently swallowing any exception.

Passing a string, as you proposed, may not be enough, imo, as we will lose some information along the way. Passing the whole `Exception` object, otoh, should be enough for the client code to decide what to do next.

P.S. This passage + a PR description piece could serve as a source of the future documentation for #74. =)

Reply:
Thank you for the explanations on the thought process.

I like how this is laid out now; the response map can easily be checked for the existence of an :error keyword in all scenarios.